p2p: give seednodes time before falling back to fixed seeds

mzumsande opened the pull request:
`-seednode` is an alternative bootstrap mechanism - when choosing it, we make a `AddrFetch` connection to the specified peer, gather addresses from them, and then disconnect. Presumably, if users specify a seednode they prefer addresses from that node over fixed seeds.

However, when disabling dns seeds and specifiying `-seednode`, `CConnman::ProcessAddrFetch()`  immediately removes the entry from `m_addr_fetches` (before the seednode could give us addresses) - and once `m_addr_fetches`  is empty, `ThreadOpenConnections` will add fixed seeds, resulting in a "race" between the fixed seeds and seednodes filling up AddrMan.

This PR suggests to check for any provided `-seednode` arg instead of using the size of `m_addr_fetches`, thus delaying the querying of fixed seeds for 1 minute when specifying any seednode (as we already do for `addnode` peers).
That way, we actually give the seednodes a chance for  to provide us with addresses before falling back to fixed seeds.

This can be tested with `bitcoind -debug=net -dnsseed=0 -seednode=(...)` on a node without `peers.dat` and observing the debug log.

mzumsande committed:
net: Give seednodes time before falling back to fixed seeds

Before, we'd remove a seednode from the list right after connecting
to it, leading to a race with loading the fixed seed and connecting
to them.

sr-gi reviewed (commented):
tACK [01c7fb6](https://github.com/bitcoin/bitcoin/pull/27577/commits/01c7fb639f72f12e3d1ed3aa3da9b86ce1279e81)

I think it makes sense to prioritize nodes provided via `seednode` over fixed seeds. However, this makes me think whether we'd like to do the same over `dnsseed`.

The proposed changes LGTM. Replacing the `m_addr_fetch.empty()` by `use_seednodes` doesn't seem to have any side effects, given `m_addr_fetch` is only populated by addresses provided via `seednode` and by `ThreadDNSAddressSeed`. In the former, it means that it cannot be the case that `m_addr_fetch` is empty if `use_seednodes` is set. Whereas for the latter, this only triggers provided `dnsseed` is enabled, so it won't matterfor our case.

There are a couple of things I realized that may be simplifiable even though they are not directly modified by this PR, but are part of the logic around it, check comments inline.

ajtowns commented on src/net.cpp:
```diff
@@ -1670,12 +1671,12 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
             }
 
-            // Checking !dnsseed is cheaper before locking 2 mutexes.
+            // Checking !dnsseed is cheaper before locking a mutex.
```
Checking `use_seednodes` is cheap too, so should also be checked before acquiring the mutex rather than after.

mzumsande replied:
fixed

ajtowns reviewed (commented):
Approach ACK 01c7fb639f72f12e3d1ed3aa3da9b86ce1279e81

mzumsande commented on src/net.cpp:
```diff
@@ -1670,12 +1671,12 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
 
-            // Checking !dnsseed is cheaper before locking 2 mutexes.
+            // Checking !dnsseed is cheaper before locking a mutex.
             if (!add_fixed_seeds_now && !dnsseed) {
```
good suggestion, I took it, that should make the code easier to read.

mzumsande commented on src/net.cpp:
```diff
@@ -1641,6 +1641,7 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
     auto next_extra_block_relay = GetExponentialRand(start, EXTRA_BLOCK_RELAY_ONLY_PEER_INTERVAL);
     const bool dnsseed = gArgs.GetBoolArg("-dnsseed", DEFAULT_DNSSEED);
     bool add_fixed_seeds = gArgs.GetBoolArg("-fixedseeds", DEFAULT_FIXEDSEEDS);
+    const bool use_seednodes{gArgs.IsArgSet("-seednode")};
```
I think I'd prefer to leave `add_fixed_seeds_now` within the current scope, especially because my plan is to follow up by moving the entire fixed seed logic out of `ThreadOpenConnections` into what is currently `ThreadDNSAddressSeed` (currently #26114). One reason for opening this PR is so that the move into another thread can be a refactor and doesn't change timing behavior.

mzumsande commented:
[01c7fb6 ](https://github.com/bitcoin/bitcoin/commit/01c7fb639f72f12e3d1ed3aa3da9b86ce1279e81)to [3077812](https://github.com/bitcoin/bitcoin/commit/30778124b82791abdc6e930373460ef1dd587cb2): Addressed suggestions by @sr-gi  and @ajtowns

> However, this makes me think whether we'd like to do the same over dnsseed.

I agree that it could make sense to not query `-dnsseeds` if the user specified a `-seednode`, so that we don't do both at the same time - whether that's by waiting with querying the dns seeds, or soft-setting `-dnsseed` to 0 I'm not sure. But that should be a seperate PR imo.

dergoegge reviewed (approved):
Code review ACK 30778124b82791abdc6e930373460ef1dd587cb2

sr-gi commented on src/net.cpp:
```diff
@@ -1670,12 +1671,12 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
                 LogPrintf("Adding fixed seeds as 60 seconds have passed and addrman is empty for at least one reachable network\n");
             }
 
```
nit: maybe remove this line between the if/else?

mzumsande replied:
I think I'll remove it in a follow-up #26114 unless I need retouch for another reason.

sr-gi reviewed (commented):
ACK [3077812](https://github.com/bitcoin/bitcoin/pull/27577/commits/30778124b82791abdc6e930373460ef1dd587cb2) with a tiny nit, feel free to ignore it

sr-gi commented:
> [01c7fb6 ](https://github.com/bitcoin/bitcoin/commit/01c7fb639f72f12e3d1ed3aa3da9b86ce1279e81)to [3077812](https://github.com/bitcoin/bitcoin/commit/30778124b82791abdc6e930373460ef1dd587cb2): Addressed suggestions by @sr-gi and @ajtowns
> 
> > However, this makes me think whether we'd like to do the same over dnsseed.
> 
> I agree that it could make sense to not query `-dnsseeds` if the user specified a `-seednode`, so that we don't do both at the same time - whether that's by waiting with querying the dns seeds, or soft-setting `-dnsseed` to 0 I'm not sure. But that should be a seperate PR imo.

Agreed that this should be part of a separate PR.

ajtowns commented:
utACK 30778124b82791abdc6e930373460ef1dd587cb2

achow101 commented:
ACK 30778124b82791abdc6e930373460ef1dd587cb2

Code looks good and a manual test showed that there is now a 1 minute wait after attempting a seednode before the fixed seeds are added.
